Improvements and fixes in the SQL trace listener functionality

OndroMih opened the pull request:
This fixes a few issues and improves the SQL trace listener feature.

Just to remind, the SQL listener feature allows to implement the `SQLTraceListener` from GlassFish API, package it into a JAR, drop the JAR to domain/lib directory, and configure a connection pool to trigger this listener when SQL is executed. This allows to execute a custom Java code which gets info about SQL executed by the connection pool.

Fixes:

* `SQLTraceRecord` received by a listener always returned `null` from `getPoolName()`

Improvements:

* add `getSqlQuery()` method to `SQLTraceRecord` - returns the executed SQL query if a query was executed (some JDBC driver methods don't execute a query)
* add `getCallingApplicationMethod()` method to `SQLTraceRecord` - returns the info about the last method in the application that triggered the query

Adds test that verify the behavior and also demonstrate usage of the SQL trace listener feature.

OndroMih committed:
SQLTraceListener: More info in the trace record

Fixed poolName.
Added applicationName, moduleName, sqlQuery.

OndroMih committed:
SQLTraceListener: Report calling class and method from application

OndroMih committed:
SQLTraceListener: Retrieve calling application class and method

OndroMih committed:
SQLTraceListener: Test

OndroMih committed:
SQLTraceListener: Use Optional for values that can be empty

avpinchuk commented on nucleus/common/glassfish-api/src/main/java/org/glassfish/api/jdbc/SQLTraceRecord.java:
```diff
@@ -194,21 +214,111 @@ public void setParams(Object[] params) {
+     * @return Stack trace element that represents a call to a server component that triggered SQL execution
+     */
+    public Optional<StackTraceElement> getCallingApplicationMethod() {
+        return Arrays.stream(Thread.currentThread().getStackTrace())
```
Why not use `StackWalker` API with the `RETAIN_CLASS_REFERENCE`?

OndroMih replied:
Thank you for the suggestion, I forgot about this API. I redesigned my solution using StackWalker and moved the computation outside of the SQLTraceListener. It means that it will be computed even if the listener doesn't need it. I think it's OK because now the computation is more efficient. It's still computed only if there is at least 1 listener.

OndroMih committed:
SQLTraceListener: Format code to pass checkstyle

OndroMih committed:
SQLTraceListener: Redesign the method SQLTraceRecord.getCallingApplicationMethod

- use StackWalker
- eagerly find application method and set it to SQLTraceRecord (only if there are some listeners)

This should be efficient enough and leads to a simpler design.

OndroMih committed:
SQLTraceListener: Add caller method to toString()

And call it from the test.

OndroMih committed:
SQLTraceListener: Use embedded Derby for tests

dmatej commented on nucleus/common/glassfish-api/src/main/java/org/glassfish/api/jdbc/SQLTraceRecord.java:
```diff
@@ -194,21 +218,107 @@ public void setParams(Object[] params) {
             }
         }
-        // TODO add poolNames and other fields of this record.
+        if (callingApplicationMethod != null) {
```
Just invert the if, so the condition would be without negation. It is easier to read.

dmatej commented on appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/connectors/ActiveResourceAdapterImpl.java:
```diff
@@ -385,17 +389,22 @@ private void setLogWriter(ManagedConnectionFactory mcf) {
         if (jcl_ != null) {
-            mcf = (ManagedConnectionFactory) jcl_.loadClass(mcfClass).getDeclaredConstructor().newInstance();
+            mcfClass = jcl_.loadClass(mcfClassName);
+//            mcf = (ManagedConnectionFactory) mcfClass.getDeclaredConstructor().newInstance();
```
Don't forget to remove the commented out code.

dmatej reviewed (commented):
Seems good, just those two comments, thanks!

OndroMih committed:
SQLTraceListener: Cleanup code base on review

OndroMih commented:
The build is failing because of some failures unrelated to the changes in this PR.
